### PR TITLE
Centralized placement ->connection mapping & placement health management

### PR DESCRIPTION
**Pull request by anarazel**
This needs some more cleanup. But it's roughly like I think things should be.   I've included the router executor conversion here, but I can split it off as well.

I think there's some additional "architectural" work needed to properly deal with relation ownership.

### CODE REVIEW

File: src/backend/distributed/utils/placement_connection.c [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+		dlist_init(&placementEntry->connectionReferences);
+		placementEntry->failed = false;
+	}
+
+
+	/*
+	 * Check whether any of the connections already associated with the
+	 * placement can be reused, or violates FOR_DML/FOR_DDL constraints.
+	 */
+	dlist_foreach(it, &placementEntry->connectionReferences)
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: Since the main goal of these functions is to make sure we always use the same connection for placement modifications it would make the code easier to read and understand if we made this explicit by having a modifyingConnection in the placementEntry, e.g.

```
bool isModification = (flags & FOR_DML) || (flags & FOR_DDL);

if (placementEntry->modifyingConnection != NULL)
{
    /* placement has been modified, must use modifyingConnection */
    if (flags & FORCE_NEW_CONNECTION)
    {
        ereport(ERROR, (errmsg("cannot open new connections after the first modification ... ")));
    }
    if (modifyingConnection->claimedExclusively)
    {
        ereport(ERROR, (errmsg("connection already in use...")));
    }
    placementConnection = placementEntry->modifyingConnection;
}
else
{
    /* no modifying connection yet, try using a reading connection */
    dlist_foreach(it, &placementEntry->connectionReferences)
    {
        ...

        if (connection && !connection->claimedExclusively)
        {
            placementConnection = connection;
        } 
    }
}

...

if (isModification)
{
    placementEntry->modifyingConnection = placementConnection;
}

return placementConnection;
```

File: src/backend/distributed/executor/multi_router_executor.c [hunk 2]
```diff
@@ -750,20 +651,32 @@ ExecuteSingleTask(QueryDesc *queryDesc, Task *task,
...
+		{
+			connectionFlags |= FOR_DML;
+		}
+
+		/*
+		 * FIXME: It's not actually correct to use only one shard placement
+		 * here for router queries involving multiple relations. We should
+		 * check that this connection is the only modifying one associated
+		 * with all the involved shards.
+		 */
```
Comment (marcocitus) [inline src/backend/distributed/executor/multi_router_executor.c]: Any thoughts on how this will work?

It might be useful to have something like `GetPlacementJoinConnection(connectionFlags, taskPlacement, task->relationShardList)`, that looks up the placementEntry for all shards in relationShardList and returns a modifying connection for any of the placements or errors out if there is more than one.

On the other hand, If we allow CTEs that modify multiple shards then this API might not be sufficient.

File: src/backend/distributed/utils/placement_connection.c [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+AssociatePlacementWithShard(ConnectionPlacementHashEntry *placementEntry,
+							ShardPlacement *placement)
+{
+	ConnectionShardHashKey shardKey;
+	ConnectionShardHashEntry *shardEntry = NULL;
+	bool found = false;
+	dlist_iter placementIter;
+
+	shardKey.shardId = placement->shardId;
+	shardEntry = hash_search(ConnectionShardHash, &shardKey, HASH_ENTER, &found);
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: We should also make sure that co-located placements use the same connection.

File: src/backend/distributed/executor/multi_router_executor.c (hunk 2 above)
Comment (anarazel) [inline src/backend/distributed/executor/multi_router_executor.c]: >Any thoughts on how this will work?

Yes, a bit. I had a prototype for it based on the older PR.  IIRC I had added a bit more information to the tasks so it was easy to list all the associated shards.

> It might be useful to have something like GetPlacementJoinConnection(connectionFlags, taskPlacement, task->relationShardList), that looks up the placementEntry for all shards in relationShardList and returns a modifying connection for any of the placements or errors out if there is more than one.

Yes, I think that's where we'll eventually have to go.

> On the other hand, If we allow CTEs that modify multiple shards then this API might not be sufficient.

Hm. I don't quite see that coming soon. You?

File: src/backend/distributed/utils/placement_connection.c (hunk 3 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: That, or we could add knowledge about fkeys and ask for a connection referencing both placements. That'd have the advantage of having more information about locks and such. But I am more leaning towards your suggestion.

Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: To record an in-person conversation Marco and I just had:
We decided to punt on this for now. The current hack of using ```ColocatedShardIntervalList(LowestShardIntervalById())``` will continue to work, and a proper solution isn't entirely trivial. Since this PR is blocking others, it seems better to tackle that issue separately.

Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: See #1098

File: src/backend/distributed/utils/placement_connection.c (hunk 1 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: I'm unconvinced. I mean I have no problem adding a modifyingConnection, but I don't think the above coding makes things simpler. Especially as we also have to deal with the userName matching.

**Comment by anarazel** [comment]
Updated version:
- Get/StartPlacementConnection now have a userName argument (NULL implying current), connections for different users aren't shared
- CheckExistingConnections() is now a helper function for StartPlacementConnection
- some FIXMEs removed
- comment and commit message cleanup

**Comment by anarazel** [comment]
@marcocitus I think this mostly looks reasonable now?  I've not introduced ->modifyingConnection, because I didn't see much benefit given that we still need to check for conflicts with users and such anyway.  But I can if you feel strongly.

**Comment by anarazel** [comment]
Pushed a revised version which continues to prohibit expanding the set of active remote transactions. I chose to not have that restriction for selects, but it'd be trivial to add.  This also fixes #1102.

File: src/backend/distributed/utils/placement_connection.c [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+
+#include "distributed/connection_management.h"
+#include "distributed/hash_helpers.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/placement_connection.h"
+#include "utils/hsearch.h"
+#include "utils/memutils.h"
+
+
+/* information about a connection reference to a placement */
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: Maybe: A connection reference is used to register that a connection has been used to read or modify a shard placement as a particular user.

File: src/backend/distributed/utils/placement_connection.c [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+	uint64 shardId;
+	uint64 placementId;
+	const char *userName;
+
+	/* the connection */
+	MultiConnection *connection;
+
+	/* information about what the connection is used for */
+	bool hadDML;
+	bool hadDDL;
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: Worth noting why we distinguish between hadDML and hadDDL (error messages).

File: src/backend/distributed/utils/placement_connection.c [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+#include "utils/memutils.h"
+
+
+/* information about a connection reference to a placement */
+typedef struct ConnectionReference
+{
+	/* identity information about the connection */
+	uint64 shardId;
+	uint64 placementId;
+	const char *userName;
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: Worth adding a note, e.g. "Stores the user as which we read/modified the placement. We must not reuse connections that were performed using a different role, since it would not have the right permissions."

**Comment by marcocitus** [comment]
> I've not introduced ->modifyingConnection, because I didn't see much benefit given that we still need to check for conflicts with users and such anyway. 

I think CheckExistingConnections is quite easy to understand now, so even though I'd have a slight preference for modifyingConnection, I could live without it. I do think it's worth adding some explicit comment(s) that there can be only one modifying connection.

File: src/backend/distributed/transaction/transaction_management.c [hunk 7]
```diff
@@ -215,6 +215,8 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
 
 		case XACT_EVENT_PRE_COMMIT:
 		{
+			bool using2PC = MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC;
```
Comment (marcocitus) [inline src/backend/distributed/transaction/transaction_management.c]: This name seems a bit misleading, we might not be using 2PC here even if MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC.

File: src/backend/distributed/transaction/transaction_management.c [hunk 8]
```diff
@@ -251,12 +268,11 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
...
-			 * Call other parts of citus that need to integrate into
-			 * transaction management. Call them *after* committing/preparing
-			 * the remote transactions, to allow marking shards as invalid
-			 * (e.g. if the remote commit failed).
+			 *
+			 * Check again whether shards/placement successfully
+			 * committed. This handles failure at COMMIT/PREPARE time.
 			 */
-			RouterExecutorPreCommitCheck();
+			CheckForFailedPlacements(false, using2PC);
```
Comment (marcocitus) [inline src/backend/distributed/transaction/transaction_management.c]: Would it make sense to just check the GUC in CheckShardPlacements? I wasn't quite sure what using2PC was referring to.

File: src/backend/distributed/utils/placement_connection.c [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+		}
+		else if (flags & FORCE_NEW_CONNECTION)
+		{
+			/* no connection reuse, just check for conflicts */
+			useConnection = false;
+		}
+		else if (returnConnectionReference != NULL)
+		{
+			/* already found connection, just check for conflicts */
+			useConnection = false;
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: Could there be a situation in which the list of connections contains 2 usable connections and the first one only read from the placement while the second one has returned DML/DDL, and the current connection is FOR_DML/FOR_DDL? Hence we error out instead of returning the modifying connection?

File: src/backend/distributed/utils/placement_connection.c [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+		}
+	}
+
+	return returnConnectionReference;
+}
+
+
+/*
+ * AssociatePlacementWithShard records shard->placement relation in
+ * ConnectionShardHash.
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: A comment on why we need this would be useful.

File: src/backend/distributed/utils/placement_connection.c [hunk 11]
```diff
@@ -0,0 +1,574 @@
...
+ * invalid, and/or whether sufficiently many placements have failed to abort
+ * the entire coordinated transaction.
+ *
+ * This will usually be called twice. Once before the remote commit is done,
+ * and once after. This is so we can abort before executing remote commits,
+ * and so we can handle remote transactions that failed during commit.
+ *
+ * When preCommit or using2PC is true, failures on transactions marked as
+ * critical will abort the entire coordinated transaction. Otherwise we can't
+ * anymore, because some remote transactions might have already committed.
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: "can't anymore" -> "can't roll back anymore"

File: src/backend/distributed/utils/placement_connection.c [hunk 12]
```diff
@@ -0,0 +1,574 @@
...
+		}
+		else
+		{
+			elevel = WARNING;
+		}
+
+		ereport(elevel,
+				(errmsg("could not commit transaction for shard " INT64_FORMAT
+						" on any active node",
+						shardEntry->key.shardId)));
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: I'd prefer moving the above logic into CheckForFailedPlacements to make it clear that we decide to continue with the rest of the loop in CheckForFailedPlacements in the 1PC case.

File: src/backend/distributed/utils/placement_connection.c [hunk 13]
```diff
@@ -0,0 +1,574 @@
...
+	{
+		int elevel = 0;
+
+		/*
+		 * Only error out if we're pre-commit or using 2PC. Otherwise we can
+		 * end up with a state where parts of the transaction is committed and
+		 * others aren't, without correspondingly marking things as invalid
+		 * (which we can't, as we would have already committed).  This sucks.
+		 */
+		if (preCommit || using2PC)
```
Comment (marcocitus) [inline src/backend/distributed/utils/placement_connection.c]: This  using2PC logic is a little confusing. In the router executor case, we would not error out here if all commits failed and MultiShardCommitProtocol is set to COMMIT_PROTOCOL_1PC?

File: src/backend/distributed/utils/placement_connection.c (hunk 9 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: I don't think that'd be ok - we should have errored out establishing the connection FOR_DML/FOR_DDL, no? Having a second active connection while the other is DML/DDL would violate read-your-own-writes/deadlock respectively. Right?

File: src/backend/distributed/utils/placement_connection.c (hunk 13 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: We can't error out, because there might be other shards that have been [partially] modified and committed successfully :( (if the other shard had fewer placements).  If no modifications at all are successful we'll error out later in CheckForFailedPlacements().   I'll add a comment explaining that we'll fail later if no modifications were successful?

File: src/backend/distributed/utils/placement_connection.c (hunk 12 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: Hm. I'm not sure what you're envisioning? The two functions used to be one, but I thought it'd be slightly easier to understand them this way. But maybe it's better after all to merge them?

File: src/backend/distributed/transaction/transaction_management.c (hunk 7 above)
Comment (anarazel) [inline src/backend/distributed/transaction/transaction_management.c]: It's definitely wrong, yea. Should be CurrentTransactionUse2PC.

File: src/backend/distributed/transaction/transaction_management.c (hunk 8 above)
Comment (anarazel) [inline src/backend/distributed/transaction/transaction_management.c]: The guc usage is wrong, a mistake in my rebase. It should just be CurrentTransactionUse2PC...

File: src/backend/distributed/utils/placement_connection.c (hunk 11 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: done, plus some minor additional rephrasing

File: src/backend/distributed/utils/placement_connection.c (hunk 10 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: done

File: src/backend/distributed/utils/placement_connection.c (hunk 5 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: done

File: src/backend/distributed/utils/placement_connection.c (hunk 6 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: done

File: src/backend/distributed/utils/placement_connection.c (hunk 4 above)
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: done

File: src/backend/distributed/utils/placement_connection.c [hunk 14]
```diff
@@ -0,0 +1,574 @@
+/*-------------------------------------------------------------------------
+ *
+ * placement_connection.c
```
Comment (anarazel) [inline src/backend/distributed/utils/placement_connection.c]: done

**Comment by marcocitus** [comment]
> I don't think that'd be ok - we should have errored out establishing the connection FOR_DML/FOR_DDL, no? Having a second active connection while the other is DML/DDL would violate read-your-own-writes/deadlock respectively. Right?

A FOR_DML command starting over a separate connection after a SELECT seems fine. This could be the case when a connection that is used for a SELECT is claimed exclusively (e.g. cursors) and then a DML is performed on another connection. If the first connection is unclaimed, we can support another write on the second connection, but at the moment it would error out. If a SELECT follows DML/DDL it should always default to the modifying connection.

A FOR_DDL command starting after a SELECT command over a separate connection should be forbidden, but we don't seem to error out in that case...

I'm again thinking that a modifyingConnection variable would make the logic in CheckExistingConnections far easier to understand and make correct.

**Comment by anarazel** [comment]
> A FOR_DML command starting over a separate connection after a SELECT seems fine. This could be the case when a connection that is used for a SELECT is claimed exclusively (e.g. cursors) and then a DML is performed on another connection. If the first connection is unclaimed, we can support another write on the modifying connection, but at the moment it would error out. If a SELECT follows DML/DDL it should also default to the modifying connection.

I don't think that's ok. We'd violate read-your-own-writes if the select question does multiple statements.  I'm not saying that there aren't situations where it'd be useful to have select connections and a DML one (in that order), but it seems quite failure prone.  I'd rather not allow it for now.

> A FOR_DDL command starting after a SELECT command over a separate connection should be forbidden, but we don't seem to error out in that case...

Oops, that' should be done.

A FOR_DML command starting over a separate connection after a SELECT seems fine. This could be the case when a connection that is used for a SELECT is claimed exclusively (e.g. cursors) and then a DML is performed on another connection. If the first connection is unclaimed, we can support another write on the modifying connection, but at the moment it would error out. If a SELECT follows DML/DDL it should also default to the modifying connection.

A FOR_DDL command starting after a SELECT command over a separate connection should be forbidden, but we don't seem to error out in that case...

> I'm again thinking that a modifyingConnection variable would make the logic in CheckExistingConnections far easier to understand and make correct.

Think it'll be more verbose, without much gain, but I'll try.

File: src/backend/distributed/connection/placement_connection.c [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+	 * possible. That's because we only can execute DML/DDL over that
+	 * connection.  Checking this first also allows us to detect conflicts due
+	 * to opening a second connection that wants to execute DML/DDL.
+	 */
+	if (placementEntry->modifyingConnection)
+	{
+		ConnectionReference *connectionReference = placementEntry->modifyingConnection;
+		MultiConnection *connection = connectionReference->connection;
+
+		if (!connection)
```
Comment (marcocitus) [inline src/backend/distributed/connection/placement_connection.c]: It seems like we can capture all the if-blocks in something like `if (CanUseExistingConnection(connectionReference, flags, userName)) return connectionReference;` and reuse it in the for loop below.

File: src/backend/distributed/executor/multi_router_executor.c [hunk 16]
```diff
@@ -1593,7 +1416,7 @@ ConsumeQueryResult(PGconn *connection, bool failOnError, int64 *rows)
 	 */
 	while (true)
 	{
-		PGresult *result = PQgetResult(connection);
+		PGresult *result = PQgetResult(connection->pgConn);
```
Comment (marcocitus) [inline src/backend/distributed/executor/multi_router_executor.c]: Since we're changing this, is it possible to call GetRemoteCommandResult here or would you rather do that as a separate PR?

File: src/backend/distributed/connection/placement_connection.c [hunk 17]
```diff
@@ -0,0 +1,652 @@
...
+							"executed on existing placement connection")));
+		}
+		else if (connectionReference->hadDML)
+		{
+			/* we'd not see the other connection's contents */
+			ereport(ERROR,
+					(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+					 errmsg("cannot establish new placement connection when DML has been "
+							"executed on existing placement connection")));
+		}
```
Comment (marcocitus) [inline src/backend/distributed/connection/placement_connection.c]: I guess either hadDML or hadDDL should always be true, right? i.e. we should always error out. Maybe we can add an an assert so readers know we shouldn't reach the code below.

File: src/backend/distributed/executor/multi_router_executor.c (hunk 16 above)
Comment (anarazel) [inline src/backend/distributed/executor/multi_router_executor.c]: I'd rather do it in a separate PR.

File: src/backend/distributed/connection/placement_connection.c (hunk 17 above)
Comment (anarazel) [inline src/backend/distributed/connection/placement_connection.c]: wfm

File: src/backend/distributed/connection/placement_connection.c (hunk 15 above)
Comment (anarazel) [inline src/backend/distributed/connection/placement_connection.c]: ok
